### PR TITLE
Support atexit() in standalone mode. helps #10990

### LINKED ISSUE #10990: Stdio streams are not flushed in standalone mode
To reproduce with the last version (`1.39.13`):

```
$ upstream/emscripten/emcc hello.c -s EXIT_RUNTIME=1 -o hello.wasm
$ wasmer hello.wasm
```

With `hello.c` containing:

```c
#include <stdio.h>
int main() { printf("Hello World"); return 0; }
```

---

In JS, this is handled by `flush_NO_FILESYSTEM`, which is in the `__ATEXIT__` array:

```js
function flush_NO_FILESYSTEM() {
    // flush anything remaining in the buffers during shutdown
    if (typeof _fflush !== 'undefined') _fflush(0);
    var buffers = SYSCALLS.buffers;
    if (buffers[1].length) SYSCALLS.printChar(1, 10);
    if (buffers[2].length) SYSCALLS.printChar(2, 10);
}
```

Would it be possible to generate something similar at the end of `_start` when `EXIT_RUNTIME` and `STANDALONE_WASM` are set?

### PR DESCRIPTION
**Pull request by kripken**
@sbc100 I can use some help with the weak linking here, I seem
to be doing it wrong. What I want is for `atexit.c` to only be linked in
when it is needed, and for `crt1.c` to have it be a null pointer if it
isn't linked in. That works in all cases except for
`tests/other/metadce/mem_no_argv_O3_STANDALONE_WASM_flto.funcs`
where the diff here shows `__funcs_on_exit` from `atexit.c`. Any idea
why things are different in LTO mode?

### CODE REVIEW

**Comment by kripken** [comment]
Oh, actually, it looks like musl doesn't actually use `atexit` to flush
the streams. It does it manually in its `exit()` impl.

I switched this to use that, but I'm not sure it's a good idea as it's
larger?

**Comment by sbc100** [comment]
Looks good!

Will this break people who are using EXIT_RUNTIME=0?   I'm have some chagnes planed to allow standalone programs without a main (i.e. reactors).   In this future, EXIT_RUNTIME won't make any sense for standalone mode since either you have a main (and the runtime exists) or you are a reactor without a main (and the runtime keeps going).    There should no case for having a main and also being a reactor.

**Comment by sbc100** [comment]
We should probably be tracking the size of our standalone output.. if the binaries are larger I would expect to see expectations change, no?

**Comment by kripken** [comment]
This shouldn't break anyone, unless they expected atexits to not happen, and stdio not to be flushed - so seems safe to me?

We sort of track code size - with a little +-, since it changes due to LLVM all the time. This change doesn't make a big enough difference to notice in this case, it seems. But the new functions show more code is added that we don't want, as the expectations files show.

I'm at a loss for how to figure out what to do to fix that. I tried all sorts of things but it always ends up always linking in parts of libc unnecessarily, and I don't know why. Is there a way to tell the linker "tell me why you linked in `__stdio_exit.o`" for example? It gets linked in even with a program that does no stdio at all, and I see no reason for it.

**Comment by kripken** [comment]
(it gets linked in, then it places itself in the table, and at that point no further DCE can remove it)

**Comment by sbc100** [comment]
You can add `--trace-symbol=foo` to the link flags to see why a given symbol is included.

File: tests/core/test_atexit_threads.c [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/*
```
Comment (sbc100) [inline tests/core/test_atexit_threads.c:1]: I find the name of this test confusing since it doesn't relate to threading.

Comment (sbc100) [reply tests/core/test_atexit_threads.c:1]: how about `test_atexit_cxa`

**Comment by kripken** [comment]
> You can add --trace-symbol=foo to the link flags to see why a given symbol is included.

Nice, I wasn't aware of that!

Ok, this helps, but I've gotten stuck and could use help:

```
$ ./emcc a.cpp -s STANDALONE_WASM -O3 --profiling
.emscripten_cache/wasm/libc.a: lazy definition of __stdio_exit_needed
.emscripten_cache/wasm/libc_rt_wasm.a(__towrite.c.o): reference to __stdio_exit_needed
.emscripten_cache/wasm/libc.a(__stdio_exit.c.o): definition of __stdio_exit_needed

$ ./emcc a.cpp -s STANDALONE_WASM -O3 --profiling
.emscripten_cache/wasm/libc_rt_wasm.a: lazy definition of __towrite
.emscripten_cache/wasm/libc_rt_wasm.a(fwrite.c.o): reference to __towrite
.emscripten_cache/wasm/libc_rt_wasm.a(__towrite.c.o): definition of __towrite

$ ./emcc a.cpp -s STANDALONE_WASM -O3 --profiling
.emscripten_cache/wasm/libc_rt_wasm.a: lazy definition of fwrite
.emscripten_cache/wasm/libc_rt_wasm.a(fwrite.c.o): definition of fwrite
```

My guess as to what this means:

* `__towrite` calls `__stdio_exit`
* `fwrite` calls `__towrite`

But what calls `fwrite`? That last invocation doesn't seem to explain why `fwrite.c.o` was linked in, which seems to be the problem...

**Comment by sbc100** [comment]
What is your `a.cpp` program in this case?

**Comment by kripken** [comment]
It's empty :smiley: 
```cpp
int main() {
}
```
Nothing there at all...

**Comment by kripken** [comment]
Ok, this should be good to go now after that malloc fix landed, and I rewrote it.

@sbc100 PTAL at the `exit/_exit/_Exit` etc changes here, I hope I did those right...

**Comment by kripken** [comment]
We could have an ifdef perhaps, would you prefer that?

The issue is that we don't support threads in standalone mode yet, so we don't want to test `cxa_thread_*`.

**Comment by sbc100** [comment]
> We could have an ifdef perhaps, would you prefer that?
> 
> The issue is that we don't support threads in standalone mode yet, so we don't want to test `cxa_thread_*`.

Ah..  I had misunderstood what this function was for.  Its for destructors of TLS variable.

I think we should look into this some more.   But perhaps that can happens a followup.  For example it should be possible to declare use thread_local variables even in single threaded programs, no?

Also it seems trance that we are testing __cxa_thread_atexit_impl which seems like an internal symbols.

File: tests/core/test_atexit_cxa.c [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/*
```
Comment (sbc100) [inline tests/core/test_atexit_cxa.c:1]: Sorry can you rename this back to test_atexit_tread?   I'm sorry I misunderstood what `__cxa_thread_atexit` was for.

With that lgtm.

**Pull request merged**